src/PowerShellEditorServices/Language/TokenOperations.cs

```diff
@@ -0,0 +1,271 @@
+            // first. This is needed to detect duplicate regions. The first in the list
+            // will be used and subsequent duplicates ignored.
+            foldableRegions.Sort();
+
+            // It's possible to have duplicate or overlapping ranges, that is, regions which have the same starting
+            // line number as the previous region. Therefore only emit ranges which have a different starting line
+            // than the previous range.
+            foldableRegions.RemoveAll( (FoldingReference item) => {
+                // Note - I'm not happy with searching here, but as the RemoveAll
+                // doesn't expose the index in the List, we need to calculate it. Fortunately the
+                // list is sorted at this point, so we can use BinarySearch.
+                int index = foldableRegions.BinarySearch(item);
+                if (index == 0) { return false; }
+                return (item.StartLine == foldableRegions[index - 1].StartLine);
+            });
+
+            // Some editors have different folding UI, sometimes the lastline should be displayed
+            // If we do want to show the last line, just change the region to be one line less
+            if (ShowLastLine) {
+                foldableRegions.ForEach( item => { item.EndLine--; });
+            }
+
+            return foldableRegions.ToArray();
+        }
+
+        /// <summary>
+        /// Creates an instance of a FoldingReference object from a start and end langauge Token
+        /// Returns null if the line range is invalid
+        /// </summary>
+        static private FoldingReference CreateFoldingReference(
+            Token startToken,
+            Token endToken,
+            string matchKind)
+        {
+            if (endToken.Extent.EndLineNumber == startToken.Extent.StartLineNumber) { return null; }
+            // Extents are base 1, but LSP is base 0, so minus 1 off all lines and character positions
+            return new FoldingReference(
+                startToken.Extent.StartLineNumber - 1,
+                startToken.Extent.StartColumnNumber - 1,
+                endToken.Extent.EndLineNumber - 1,
+                endToken.Extent.EndColumnNumber - 1,
+                matchKind
+            );
+        }
+
+        /// <summary>
+        /// Creates an instance of a FoldingReference object from a start token and an end line
+        /// Returns null if the line range is invalid
+        /// </summary>
+        static private FoldingReference CreateFoldingReference(
+            Token startToken,
+            int endLine,
+            string matchKind)
+        {
+            if (endLine == (startToken.Extent.StartLineNumber - 1)) { return null; }
+            return new FoldingReference(
+                startToken.Extent.StartLineNumber - 1,   // Extents are base 1, but LSP is base 0
+                startToken.Extent.StartColumnNumber - 1, // Extents are base 1, but LSP is base 0
```

Comment:
can you put this as one comment above the return?

Reply:
Fixed